File: DNA-Based Biomarker Discovery and Computational Drug Target Identification/scripts/dna_ml_pipeline.py
"""
Machine Learning Pipeline for DNA-Based Biomarker Discovery

This script trains machine learning models to identify potential biomarkers
and drug targets from DNA sequence features.

Author: DNA-Based Biomarker Discovery Project
Date: 2025
"""

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import (
    classification_report, confusion_matrix, 
    roc_auc_score, roc_curve, accuracy_score
)
from sklearn.preprocessing import StandardScaler
import joblib
import argparse
import os
import sys

# Import feature extraction function
sys.path.append(os.path.dirname(os.path.abspath(__file__)))
from dna_feature_extraction import extract_dna_features, get_feature_names


def train_models(X_train, y_train, X_test, y_test, use_scaling=True):
    """
    Train multiple machine learning models and evaluate their performance.
    
    Parameters:
    -----------
    X_train : np.ndarray
        Training feature matrix
    y_train : np.ndarray
        Training labels
    X_test : np.ndarray
        Test feature matrix
    y_test : np.ndarray
        Test labels
    use_scaling : bool
        Whether to scale features for models that require it
        
    Returns:
    --------
    dict
        Dictionary containing trained models and their performance metrics
    """
    # Scale features if needed
    scaler = None
    if use_scaling:
        scaler = StandardScaler()
        X_train_scaled = scaler.fit_transform(X_train)
        X_test_scaled = scaler.transform(X_test)
    
    # Define models
    models = {
        'Random Forest': RandomForestClassifier(
            n_estimators=100, 
            random_state=42, 
            n_jobs=-1,
            max_depth=10
        ),
        'Gradient Boosting': GradientBoostingClassifier(
            n_estimators=100, 
            random_state=42,
            max_depth=5
        ),
        'SVM': SVC(
            kernel='rbf', 
            probability=True, 
            random_state=42,
            C=1.0,
            gamma='scale'
        ),
        'Logistic Regression': LogisticRegression(
            random_state=42, 
            max_iter=1000, 
            n_jobs=-1,
            C=1.0
        )
    }
    
    results = {}
    
    for name, model in models.items():
        print(f"\n{'='*50}")
        print(f"Training {name}...")
        print(f"{'='*50}")
        
        # Use scaled data for SVM and Logistic Regression
        if name in ['SVM', 'Logistic Regression'] and use_scaling:
            X_train_use = X_train_scaled
            X_test_use = X_test_scaled
        else:
            X_train_use = X_train
            X_test_use = X_test
        
        # Train model
        model.fit(X_train_use, y_train)
        
        # Predictions
        y_pred = model.predict(X_test_use)
        y_pred_proba = model.predict_proba(X_test_use)[:, 1]
        
        # Metrics
        accuracy = accuracy_score(y_test, y_pred)
        auc_score = roc_auc_score(y_test, y_pred_proba)
        
        # Cross-validation
        cv_scores = cross_val_score(
            model, X_train_use, y_train, 
            cv=5, scoring='accuracy', n_jobs=-1
        )
        
        results[name] = {
            'model': model,
            'accuracy': accuracy,
            'auc': auc_score,
            'cv_mean': cv_scores.mean(),
            'cv_std': cv_scores.std(),
            'y_pred': y_pred,
            'y_pred_proba': y_pred_proba
        }
        
        print(f"Accuracy: {accuracy:.4f}")
        print(f"AUC-ROC: {auc_score:.4f}")
        print(f"CV Accuracy: {cv_scores.mean():.4f} (+/- {cv_scores.std() * 2:.4f})")
        print(f"\nClassification Report:")
        print(classification_report(y_test, y_pred))
    
    return results, scaler


def get_feature_importance(model, feature_names):
    """
    Get feature importance from a tree-based model.
    
    Parameters:
    -----------
    model : sklearn model
        Trained model with feature_importances_ attribute
    feature_names : List[str]
        List of feature names
        
    Returns:
    --------
    pd.DataFrame
        DataFrame with features and their importance scores
    """
    if hasattr(model, 'feature_importances_'):
        importance_df = pd.DataFrame({
            'feature': feature_names,
            'importance': model.feature_importances_
        }).sort_values('importance', ascending=False)
        return importance_df
    else:
        return None


def main():
    """Main function to run the ML pipeline."""
    parser = argparse.ArgumentParser(
        description='Train ML models for DNA biomarker discovery'
    )
    parser.add_argument(
        '--input', 
        type=str, 
        default='../data/genomics_data.csv',
        help='Path to input CSV file with DNA sequences'
    )
    parser.add_argument(
        '--features', 
        type=str, 
        default=None,
        help='Path to pre-extracted features CSV (optional)'
    )
    parser.add_argument(
        '--output-dir', 
        type=str, 
        default='../results/models',
        help='Directory to save trained models'
    )
    parser.add_argument(
        '--test-size',
        type=float,
        default=0.2,
        help='Proportion of data to use for testing'
    )
    parser.add_argument(
        '--random-state',
        type=int,
        default=42,
        help='Random state for reproducibility'
    )
    
    args = parser.parse_args()
    
    # Load data
    print(f"Loading data from {args.input}...")
    df = pd.read_csv(args.input)
    
    # Extract or load features
    if args.features and os.path.exists(args.features):
        print(f"Loading pre-extracted features from {args.features}...")
        features_df = pd.read_csv(args.features)
        if 'Labels' in features_df.columns:
            X = features_df.drop('Labels', axis=1).values
            y = features_df['Labels'].values
        else:
            X = features_df.values
            y = df['Labels'].values
        feature_names = features_df.columns.tolist()
        if 'Labels' in feature_names:
            feature_names.remove('Labels')
    else:
        print("Extracting features from DNA sequences...")
        sequences = df['Sequences'].values.tolist()
        X = extract_dna_features(sequences)
        y = df['Labels'].values
        feature_names = get_feature_names()
    
    print(f"Feature matrix shape: {X.shape}")
    print(f"Number of features: {X.shape[1]}")
    
    # Split data
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, 
        test_size=args.test_size, 
        random_state=args.random_state, 
        stratify=y
    )
    
    print(f"\nTraining set size: {X_train.shape[0]}")
    print(f"Test set size: {X_test.shape[0]}")
    
    # Train models
    results, scaler = train_models(X_train, y_train, X_test, y_test)
    
    # Find best model
    best_model_name = max(results.keys(), key=lambda x: results[x]['auc'])
    best_model = results[best_model_name]['model']
    
    print(f"\n{'='*50}")
    print(f"Best model: {best_model_name}")
    print(f"Best AUC-ROC: {results[best_model_name]['auc']:.4f}")
    print(f"Best accuracy: {results[best_model_name]['accuracy']:.4f}")
    print(f"{'='*50}")
    
    # Get feature importance
    if hasattr(best_model, 'feature_importances_'):
        importance_df = get_feature_importance(best_model, feature_names)
        print(f"\nTop 10 Most Important Features:")
        print(importance_df.head(10))
        
        # Save feature importance
        importance_path = os.path.join(args.output_dir, 'feature_importance.csv')
        os.makedirs(args.output_dir, exist_ok=True)
        importance_df.to_csv(importance_path, index=False)
        print(f"\nFeature importance saved to {importance_path}")
    
    # Save models
    os.makedirs(args.output_dir, exist_ok=True)
    
    # Save best model
    best_model_path = os.path.join(
        args.output_dir, 
        f'best_model_{best_model_name.lower().replace(" ", "_")}.pkl'
    )
    joblib.dump(best_model, best_model_path)
    print(f"Best model saved to {best_model_path}")
    
    # Save scaler if used
    if scaler is not None:
        scaler_path = os.path.join(args.output_dir, 'scaler.pkl')
        joblib.dump(scaler, scaler_path)
        print(f"Scaler saved to {scaler_path}")
    
    # Save feature names
    feature_names_path = os.path.join(args.output_dir, 'feature_names.pkl')
    joblib.dump(feature_names, feature_names_path)
    print(f"Feature names saved to {feature_names_path}")
    
    # Save results summary
    results_summary = pd.DataFrame({
        'Model': list(results.keys()),
        'Accuracy': [results[m]['accuracy'] for m in results.keys()],
        'AUC-ROC': [results[m]['auc'] for m in results.keys()],
        'CV Accuracy': [results[m]['cv_mean'] for m in results.keys()],
        'CV Std': [results[m]['cv_std'] for m in results.keys()]
    })
    results_path = os.path.join(args.output_dir, 'model_comparison.csv')
    results_summary.to_csv(results_path, index=False)
    print(f"Results summary saved to {results_path}")
    
    print("\nPipeline complete!")


if __name__ == "__main__":
    main()

File: DNA-Based Biomarker Discovery and Computational Drug Target Identification/scripts/dna_ml_pipeline.R
# Machine Learning Pipeline for DNA-Based Biomarker Discovery
#
# This script trains machine learning models to identify potential biomarkers
# and drug targets from DNA sequence features.
#
# Author: DNA-Based Biomarker Discovery Project
# Date: 2025

# Load required libraries
library(caret)
library(randomForest)
library(e1071)
library(pROC)
library(dplyr)
library(data.table)

# Source feature extraction script
source("dna_feature_extraction.R")

#' Train multiple machine learning models and evaluate their performance
#'
#' @param X_train Training feature matrix
#' @param y_train Training labels
#' @param X_test Test feature matrix
#' @param y_test Test labels
#' @param use_scaling Whether to scale features for models that require it
#' @return List containing trained models and their performance metrics
train_models <- function(X_train, y_train, X_test, y_test, use_scaling = TRUE) {
  results <- list()
  
  # Scale features if needed
  scaler <- NULL
  if (use_scaling) {
    preproc_params <- preProcess(X_train, method = c("center", "scale"))
    X_train_scaled <- predict(preproc_params, X_train)
    X_test_scaled <- predict(preproc_params, X_test)
    scaler <- preproc_params
  }
  
  # Convert labels to factors
  y_train_factor <- as.factor(y_train)
  y_test_factor <- as.factor(y_test)
  
  # 1. Random Forest
  cat("\n", rep("=", 50), "\n", sep = "")
  cat("Training Random Forest...\n")
  cat(rep("=", 50), "\n", sep = "")
  
  rf_model <- randomForest(
    x = X_train,
    y = y_train_factor,
    ntree = 100,
    mtry = sqrt(ncol(X_train)),
    importance = TRUE,
    do.trace = FALSE
  )
  
  rf_pred <- predict(rf_model, X_test, type = "response")
  rf_pred_proba <- predict(rf_model, X_test, type = "prob")[, 2]
  
  rf_accuracy <- mean(rf_pred == y_test_factor)
  rf_auc <- as.numeric(auc(roc(y_test, rf_pred_proba)))
  
  # Cross-validation for Random Forest
  rf_cv <- train(
    x = X_train,
    y = y_train_factor,
    method = "rf",
    trControl = trainControl(method = "cv", number = 5),
    tuneGrid = data.frame(mtry = sqrt(ncol(X_train)))
  )
  
  results$RandomForest <- list(
    model = rf_model,
    accuracy = rf_accuracy,
    auc = rf_auc,
    cv_accuracy = max(rf_cv$results$Accuracy),
    predictions = rf_pred,
    probabilities = rf_pred_proba
  )
  
  cat("Accuracy:", rf_accuracy, "\n")
  cat("AUC-ROC:", rf_auc, "\n")
  cat("CV Accuracy:", max(rf_cv$results$Accuracy), "\n")
  
  # 2. Support Vector Machine
  cat("\n", rep("=", 50), "\n", sep = "")
  cat("Training SVM...\n")
  cat(rep("=", 50), "\n", sep = "")
  
  svm_model <- svm(
    x = X_train_scaled,
    y = y_train_factor,
    kernel = "radial",
    probability = TRUE,
    cost = 1,
    gamma = 1 / ncol(X_train_scaled)
  )
  
  svm_pred <- predict(svm_model, X_test_scaled)
  svm_pred_proba <- attr(predict(svm_model, X_test_scaled, probability = TRUE), 
                         "probabilities")[, 2]
  
  svm_accuracy <- mean(svm_pred == y_test_factor)
  svm_auc <- as.numeric(auc(roc(y_test, svm_pred_proba)))
  
  # Cross-validation for SVM
  svm_cv <- train(
    x = X_train_scaled,
    y = y_train_factor,
    method = "svmRadial",
    trControl = trainControl(method = "cv", number = 5),
    tuneGrid = expand.grid(
      C = c(0.1, 1, 10),
      sigma = c(0.01, 0.1, 1)
    )
  )
  
  results$SVM <- list(
    model = svm_model,
    accuracy = svm_accuracy,
    auc = svm_auc,
    cv_accuracy = max(svm_cv$results$Accuracy),
    predictions = svm_pred,
    probabilities = svm_pred_proba
  )
  
  cat("Accuracy:", svm_accuracy, "\n")
  cat("AUC-ROC:", svm_auc, "\n")
  cat("CV Accuracy:", max(svm_cv$results$Accuracy), "\n")
  
  # 3. Logistic Regression
  cat("\n", rep("=", 50), "\n", sep = "")
  cat("Training Logistic Regression...\n")
  cat(rep("=", 50), "\n", sep = "")
  
  # Prepare data for logistic regression
  train_data <- data.frame(X_train_scaled, Labels = y_train_factor)
  test_data <- data.frame(X_test_scaled, Labels = y_test_factor)
  
  glm_model <- glm(
    Labels ~ .,
    data = train_data,
    family = binomial(link = "logit")
  )
  
  glm_pred_proba <- predict(glm_model, test_data, type = "response")
  glm_pred <- ifelse(glm_pred_proba > 0.5, 1, 0)
  glm_pred_factor <- factor(glm_pred, levels = levels(y_test_factor))
  
  glm_accuracy <- mean(glm_pred_factor == y_test_factor)
  glm_auc <- as.numeric(auc(roc(y_test, glm_pred_proba)))
  
  # Cross-validation for Logistic Regression
  glm_cv <- train(
    Labels ~ .,
    data = train_data,
    method = "glm",
    family = "binomial",
    trControl = trainControl(method = "cv", number = 5)
  )
  
  results$LogisticRegression <- list(
    model = glm_model,
    accuracy = glm_accuracy,
    auc = glm_auc,
    cv_accuracy = max(glm_cv$results$Accuracy),
    predictions = glm_pred_factor,
    probabilities = glm_pred_proba
  )
  
  cat("Accuracy:", glm_accuracy, "\n")
  cat("AUC-ROC:", glm_auc, "\n")
  cat("CV Accuracy:", max(glm_cv$results$Accuracy), "\n")
  
  # 4. Gradient Boosting (using caret's gbm)
  cat("\n", rep("=", 50), "\n", sep = "")
  cat("Training Gradient Boosting...\n")
  cat(rep("=", 50), "\n", sep = "")
  
  gbm_cv <- train(
    x = X_train,
    y = y_train_factor,
    method = "gbm",
    trControl = trainControl(method = "cv", number = 5),
    verbose = FALSE,
    tuneGrid = expand.grid(
      interaction.depth = c(3, 5),
      n.trees = c(50, 100),
      shrinkage = c(0.01, 0.1),
      n.minobsinnode = 10
    )
  )
  
  gbm_pred <- predict(gbm_cv, X_test)
  gbm_pred_proba <- predict(gbm_cv, X_test, type = "prob")[, 2]
  
  gbm_accuracy <- mean(gbm_pred == y_test_factor)
  gbm_auc <- as.numeric(auc(roc(y_test, gbm_pred_proba)))
  
  results$GradientBoosting <- list(
    model = gbm_cv,
    accuracy = gbm_accuracy,
    auc = gbm_auc,
    cv_accuracy = max(gbm_cv$results$Accuracy),
    predictions = gbm_pred,
    probabilities = gbm_pred_proba
  )
  
  cat("Accuracy:", gbm_accuracy, "\n")
  cat("AUC-ROC:", gbm_auc, "\n")
  cat("CV Accuracy:", max(gbm_cv$results$Accuracy), "\n")
  
  return(list(results = results, scaler = scaler))
}

#' Get feature importance from a random forest model
#'
#' @param model Trained random forest model
#' @param feature_names Vector of feature names
#' @return Data frame with features and their importance scores
get_feature_importance <- function(model, feature_names) {
  if (class(model)[1] == "randomForest") {
    importance_scores <- importance(model)[, "MeanDecreaseGini"]
    importance_df <- data.frame(
      feature = feature_names,
      importance = importance_scores
    ) %>%
      arrange(desc(importance))
    return(importance_df)
  } else {
    return(NULL)
  }
}

#' Main function to run the ML pipeline
main <- function() {
  # Parse command line arguments
  args <- commandArgs(trailingOnly = TRUE)
  
  if (length(args) < 1) {
    input_file <- "../data/genomics_data.csv"
  } else {
    input_file <- args[1]
  }
  
  if (length(args) < 2) {
    output_dir <- "../results/models"
  } else {
    output_dir <- args[2]
  }
  
  # Load data
  cat("Loading data from", input_file, "...\n")
  df <- read.csv(input_file, stringsAsFactors = FALSE)
  
  # Extract features
  cat("Extracting features from DNA sequences...\n")
  sequences <- df$Sequences
  features <- extract_dna_features(sequences)
  
  # Prepare data
  if ("Labels" %in% colnames(df)) {
    y <- df$Labels
    X <- features %>% select(-Labels) %>% as.matrix()
  } else {
    stop("Labels column not found in data")
  }
  
  cat("Feature matrix shape:", nrow(X), "x", ncol(X), "\n")
  cat("Number of features:", ncol(X), "\n")
  
  # Split data
  set.seed(42)
  train_indices <- createDataPartition(y, p = 0.8, list = FALSE)
  X_train <- X[train_indices, ]
  X_test <- X[-train_indices, ]
  y_train <- y[train_indices]
  y_test <- y[-train_indices]
  
  cat("\nTraining set size:", nrow(X_train), "\n")
  cat("Test set size:", nrow(X_test), "\n")
  
  # Train models
  model_results <- train_models(X_train, y_train, X_test, y_test)
  results <- model_results$results
  scaler <- model_results$scaler
  
  # Find best model
  best_model_name <- names(results)[which.max(sapply(results, function(x) x$auc))]
  best_model <- results[[best_model_name]]$model
  
  cat("\n", rep("=", 50), "\n", sep = "")
  cat("Best model:", best_model_name, "\n")
  cat("Best AUC-ROC:", results[[best_model_name]]$auc, "\n")
  cat("Best accuracy:", results[[best_model_name]]$accuracy, "\n")
  cat(rep("=", 50), "\n", sep = "")
  
  # Get feature importance for Random Forest
  if (best_model_name == "RandomForest") {
    feature_names <- colnames(X_train)
    importance_df <- get_feature_importance(best_model, feature_names)
    
    if (!is.null(importance_df)) {
      cat("\nTop 10 Most Important Features:\n")
      print(head(importance_df, 10))
      
      # Save feature importance
      importance_path <- file.path(output_dir, "feature_importance.csv")
      dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
      write.csv(importance_df, importance_path, row.names = FALSE)
      cat("\nFeature importance saved to", importance_path, "\n")
    }
  }
  
  # Save models
  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
  
  # Save best model
  best_model_path <- file.path(output_dir, 
                                paste0("best_model_", 
                                       tolower(gsub(" ", "_", best_model_name)), 
                                       ".rds"))
  saveRDS(best_model, best_model_path)
  cat("Best model saved to", best_model_path, "\n")
  
  # Save scaler if used
  if (!is.null(scaler)) {
    scaler_path <- file.path(output_dir, "scaler.rds")
    saveRDS(scaler, scaler_path)
    cat("Scaler saved to", scaler_path, "\n")
  }
  
  # Save results summary
  results_summary <- data.frame(
    Model = names(results),
    Accuracy = sapply(results, function(x) x$accuracy),
    AUC_ROC = sapply(results, function(x) x$auc),
    CV_Accuracy = sapply(results, function(x) x$cv_accuracy)
  )
  results_path <- file.path(output_dir, "model_comparison.csv")
  write.csv(results_summary, results_path, row.names = FALSE)
  cat("Results summary saved to", results_path, "\n")
  
  cat("\nPipeline complete!\n")
}

# Run main function if script is executed directly
if (!interactive()) {
  main()
}

